Introduce `Rewrite & PMD` covering S1144: Unused "private" methods should be removed

Pankraz76 opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->


apply:

- https://docs.openrewrite.org/recipes/staticanalysis/removeunusedprivatemethods
- https://docs.pmd-code.org/pmd-doc-7.1.0/pmd_rules_java_bestpractices.html#unusedprivatemethod

fix: 
- https://sonarsource.github.io/rspec/#/rspec/S1144


### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).


related to:

- https://github.com/spotbugs/spotbugs/pull/3658

owaiskazi19 commented on gradle/code-convention.yml:
```diff
@@ -0,0 +1,15 @@
+#  - org.openrewrite.staticanalysis.EmptyBlock
+#  - org.openrewrite.staticanalysis.RemoveCallsToSystemGc
+#  - org.openrewrite.staticanalysis.UnnecessaryThrows
+#  - org.openrewrite.text.EndOfLineAtEndOfFile
```
Can we move this to formatting.gradle like [removeUnusedImports](https://github.com/opensearch-project/OpenSearch/blob/main/gradle/formatting.gradle#L68)() we can have `removeUnusedPrivateMethods`?

owaiskazi19 replied:
I guess I meant to say if we move `removeUnusedPrivateMethods` to formatting.gradle then we don't need `code-convention.yml` and other related files, right?

Pankraz76 replied:
sorry for confusion, yes of course can inline as well.

Pankraz76 replied:
is it better now?

Pankraz76 replied:
lets apply SOC on this for better overview. Thanks for guidance.

owaiskazi19 commented:
@andrross @cwperks can you take a look at this change as well?

Pankraz76 commented on gradle/formatting.gradle:
```diff
@@ -57,57 +61,81 @@ import org.opensearch.gradle.BuildPlugin
+        // non-standard places
+        target '**/*.java'
 
-          removeUnusedImports()
```
rewrite has its own flaws, as nobody is perfect.
In this regards it found some leftovers.

Pankraz76 commented on gradle.properties:
```diff
@@ -13,6 +13,8 @@
 org.gradle.warning.mode=none
 org.gradle.parallel=true
+# https://github.com/openrewrite/rewrite-gradle-plugin/issues/212
+org.gradle.workers.max=7
```
this really needs to be verified. The `clean` goal helps to overcome a failing build.

Pankraz76 commented:
Thanks a lot. I have given up, as the plugin changed.

From working on a 8gb machine in 40 mins to running "endless" without maxing out CPU.

@timtebeek do you have an idea what´s changed recently, causing this performance difference? As @hazendaz has pointed out correctly, every tool, including spotbugs, has some bugs.

I will try to cover this as well like recent discovery suggested by @hazendaz.
Again thanks for the contribution making rewrite better by the day.

- https://github.com/openrewrite/rewrite-static-analysis/pull/727


> Is that because of a problem with the rewrite gradle plugin, or something related to our setup?

not sure about that, most projects have some parsing error´s upfront, until its going into running recipe mode. Thats something advanced @timtebeek need to check out further. 

Appreciate it.

Pankraz76 commented on gradle/rewrite.gradle:
```diff
@@ -0,0 +1,45 @@
+  rewrite("org.openrewrite.recipe:rewrite-static-analysis:2.15.0")
+}
+
+tasks.check.dependsOn(rewriteDryRun)
```
i think its not possible to run this no default on dev machine.

Rewrite run should be the roundup solution to fix all the issues breaking the build/quality gate, just like already achieved with spotless. 


PMD is fast enough to run locally on the other hand giving quick indication for broad errors, as some are in common. Other details recipes can not be covered of course, given the need to rewrite to run (in cloud).

Pankraz76 commented:
> Thanks @Pankraz76. I checked out this PR to try it out, but ran into a few issues. I ran:
> 
> ```
> ./gradlew rewriteRun -Dorg.gradle.jvmargs=-Xmx16G
> ```
> 
> I had to use a 16G heap otherwise I ran out of memory. It found unused private methods in 36 files. However, a subsequent compilation failed because it removed the method `StarTreeMapper::Builder::findMapperBuilderByName` even though that method is still used. I reverted that change and then compilation failed because the method `SearchAfterIT::convertSortValues` had been removed. Is this expected?
> 
> Also, I get an enormous output of warnings like the following when running the tool:
> 
> > Resolution of the configuration :modules:ingest-user-agent:jacocoAnt was attempted from a context different than the project context. Have a look at the documentation to understand why this is a problem and how it can be resolved. This behavior has been deprecated. This will fail with an error in Gradle 9.0. For more information, please refer to https://docs.gradle.org/8.14.3/userguide/viewing_debugging_dependencies.html#sub:resolving-unsafe-configuration-resolution-errors in the Gradle documentation.
> 
> Is that because of a problem with the rewrite gradle plugin, or something related to our setup?

Updated the config accordingly. Please consider retest my simply executing `./gradlew rewriteDryRun -Dorg.gradle.jvmargs=-Xmx8G` with the now clickable action in the developer guide:

<img width="528" height="107" alt="image" src="https://github.com/user-attachments/assets/83a4baf4-0238-4d2c-b080-249ed09f388b" />


This should work and sort out the smells targeted with PMD and automated with rewrite. Its just as with checkstyle and spotless there is a natural evolution in this. Could argue to keep both as each having its benefits and weaknesses, meaning for the same thing both finding stuff the other one dont making it truly a win. As rewrite takes so long PMD is a good candidate to run on dev while the big bang and mighty rewrite runs on CI. 

As @hazendaz has correctly pointed out that each tool having its benefits, making each a a win for every project, alone and of course having the biggest impact in combination. 

- https://github.com/spotbugs/spotbugs/pull/3658

hazendaz commented:
I am not involved with this user @Pankraz76.  I kindly ask this user stop using my name.

Pankraz76 commented:
Kindly request some feedback on how to proceed here, thanks. @andrross
